[FLINK-15090][build] Reverse the dependency from flink-streaming-java to flink-client

TisonKun opened the pull request:
## What is the purpose of the change

https://lists.apache.org/x/thread.html/r40139010d3fe5c5964e1f0e40f9d840eaedd52a46c8f3586fd185313@%3Cdev.flink.apache.org%3E

## Brief change log

- Reverse the dependency from flink-streaming-java to flink-client
- Extract JobGraphGenerator#addUserArtifactEntries to a flink-runtime utils. Otherwise there is a dependency from flink-streaming-java to flilnk-client to flink-optimizer. flink-streaming-java should not depend on flink-optimizer also.
- Move RemoteStreamEnvironmentTest to flink-test. Otherwise there is a dependency from flink-streaming-java to flilnk-client. It is ok and see also `RemoteEnvironmentITCase`.

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes, reverse the dependency from flink-streaming-java to flink-client)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes, RemoteStreamEnvironment that explained in FLINK-16427. StreamPlanEnvironment for adjust the new structure. though, I don't think it should be ever a `@PublicEvolving` class)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

However, we change public dependencies graph, so it's worth a *release note*.

TisonKun commented:
Compilation fails on

>08:27:36.893 [ERROR] Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.11.0:cmp (default) on project flink-streaming-java_2.11: Execution default of goal com.github.siom79.japicmp:japicmp-maven-plugin:0.11.0:cmp failed: Could not load 'org.apache.flink.client.program.ProgramInvocationException': org.apache.flink.client.program.ProgramInvocationException. Please make sure that all libraries have been added to the classpath (CLASSPATH=:/home/travis/build/TisonKun/flink/flink-annotations/target/flink-annotations-1.11-SNAPSHOT.jar:/home/travis/.m2/repository/org/slf4j/slf4j-api/1.7.15/slf4j-api-1.7.15.jar:/home/travis/.m2/repository/com/google/code/findbugs/jsr305/1.3.9/jsr305-1.3.9.jar:/home/travis/.m2/repository/junit/junit/4.12/junit-4.12.jar:/home/travis/.m2/repository/org/mockito/mockito-core/2.21.0/mockito-core-2.21.0.jar:/home/travis/.m2/repository/org/powermock/powermock-module-junit4/2.0.2/powermock-module-junit4-2.0.2.jar:/home/travis/.m2/repository/org/powermock/powermock-api-mockito2/2.0.2/powermock-api-mockito2-2.0.2.jar:/home/travis/.m2/repository/org/hamcrest/hamcrest-all/1.3/hamcrest-all-1.3.jar:/home/travis/.m2/repository/org/slf4j/slf4j-log4j12/1.7.15/slf4j-log4j12-1.7.15.jar:/home/travis/.m2/repository/log4j/log4j/1.2.17/log4j-1.2.17.jar:/home/travis/build/TisonKun/flink/tools/force-shading/target/force-shading-1.11-SNAPSHOT.jar:/home/travis/.m2/repository/commons-io/commons-io/2.4/commons-io-2.4.jar:/home/travis/build/TisonKun/flink/flink-java/target/flink-java-1.11-SNAPSHOT.jar:/home/travis/.m2/repository/org/slf4j/slf4j-api/1.7.15/slf4j-api-1.7.15.jar:/home/travis/.m2/repository/org/javassist/javassist/3.24.0-GA/javassist-3.24.0-GA.jar:/home/travis/.m2/repository/org/apache/flink/flink-shaded-jackson/2.10.1-9.0/flink-shaded-jackson-2.10.1-9.0.jar:/home/travis/.m2/repository/com/twitter/chill_2.11/0.7.6/chill_2.11-0.7.6.jar:/home/travis/build/TisonKun/flink/flink-queryable-state/flink-queryable-state-client-java/target/flink-queryable-state-client-java-1.11-SNAPSHOT.jar:/home/travis/.m2/repository/com/typesafe/akka/akka-actor_2.11/2.5.21/akka-actor_2.11-2.5.21.jar:/home/travis/.m2/repository/commons-collections/commons-collections/3.2.2/commons-collections-3.2.2.jar:/home/travis/.m2/repository/com/typesafe/akka/akka-slf4j_2.11/2.5.21/akka-slf4j_2.11-2.5.21.jar:/home/travis/build/TisonKun/flink/flink-metrics/flink-metrics-core/target/flink-metrics-core-1.11-SNAPSHOT.jar:/home/travis/.m2/repository/com/typesafe/akka/akka-protobuf_2.11/2.5.21/akka-protobuf_2.11-2.5.21.jar:/home/travis/.m2/repository/com/github/scopt/scopt_2.11/3.5.0/scopt_2.11-3.5.0.jar:/home/travis/build/TisonKun/flink/tools/force-shading/target/force-shading-1.11-SNAPSHOT.jar:/home/travis/.m2/repository/commons-cli/commons-cli/1.3.1/commons-cli-1.3.1.jar:/home/travis/.m2/repository/org/clapper/grizzled-slf4j_2.11/1.3.2/grizzled-slf4j_2.11-1.3.2.jar:/home/travis/.m2/repository/com/typesafe/ssl-config-core_2.11/0.3.7/ssl-config-core_2.11-0.3.7.jar:/home/travis/build/TisonKun/flink/flink-annotations/target/flink-annotations-1.11-SNAPSHOT.jar:/home/travis/.m2/repository/com/twitter/chill-java/0.7.6/chill-java-0.7.6.jar:/home/travis/.m2/repository/org/scala-lang/modules/scala-java8-compat_2.11/0.7.0/scala-java8-compat_2.11-0.7.0.jar:/home/travis/.m2/repository/com/typesafe/akka/akka-stream_2.11/2.5.21/akka-stream_2.11-2.5.21.jar:/home/travis/.m2/repository/org/apache/commons/commons-lang3/3.3.2/commons-lang3-3.3.2.jar:/home/travis/.m2/repository/com/esotericsoftware/minlog/minlog/1.2/minlog-1.2.jar:/home/travis/build/TisonKun/flink/flink-runtime/target/flink-runtime_2.11-1.11-SNAPSHOT.jar:/home/travis/.m2/repository/org/reactivestreams/reactive-streams/1.0.2/reactive-streams-1.0.2.jar:/home/travis/.m2/repository/org/apache/commons/commons-math3/3.5/commons-math3-3.5.jar:/home/travis/.m2/repository/org/apache/flink/flink-shaded-guava/18.0-9.0/flink-shaded-guava-18.0-9.0.jar:/home/travis/.m2/repository/org/scala-lang/modules/scala-parser-combinators_2.11/1.1.1/scala-parser-combinators_2.11-1.1.1.jar:/home/travis/build/TisonKun/flink/flink-filesystems/flink-hadoop-fs/target/flink-hadoop-fs-1.11-SNAPSHOT.jar:/home/travis/.m2/repository/com/typesafe/config/1.3.0/config-1.3.0.jar:/home/travis/build/TisonKun/flink/flink-core/target/flink-core-1.11-SNAPSHOT.jar:/home/travis/.m2/repository/org/objenesis/objenesis/2.1/objenesis-2.1.jar:/home/travis/.m2/repository/org/xerial/snappy/snappy-java/1.1.4/snappy-java-1.1.4.jar:/home/travis/.m2/repository/org/apache/flink/flink-shaded-asm-7/7.1-9.0/flink-shaded-asm-7-7.1-9.0.jar:/home/travis/.m2/repository/com/esotericsoftware/kryo/kryo/2.24.0/kryo-2.24.0.jar:/home/travis/.m2/repository/com/google/code/findbugs/jsr305/1.3.9/jsr305-1.3.9.jar:/home/travis/.m2/repository/org/apache/flink/flink-shaded-netty/4.1.39.Final-9.0/flink-shaded-netty-4.1.39.Final-9.0.jar:/home/travis/.m2/repository/org/apache/commons/commons-compress/1.18/commons-compress-1.18.jar:/home/travis/.m2/repository/org/scala-lang/scala-library/2.11.12/scala-library-2.11.12.jar:) or try the option '--ignore-missing-classes'. -> [Help 1]

I have no idea how to suppress this japicmp issue since we reverse the dependency so that it can never load `ProgramAbortException` from flink-streaming-java that is in flink-client

cc @zentol @aljoscha

TisonKun commented:
The only usage of `ProgramInvocationException` in japicmp checking is

```
---  REMOVED METHOD: PUBLIC(-) org.apache.flink.api.common.JobExecutionResult execute(java.lang.String)
		---  REMOVED EXCEPTION: org.apache.flink.client.program.ProgramInvocationException
```

I think if we do this in 1.11 and mark 1.10 as the base version, conflict could be resolved.

TisonKun commented:
@zentol this is a reviewable PR for the purpose reverse dependency from flink-streaming-java to flink-client. Since dependency changes though I resolve issues on compilation there is possible some NOTICE issues, could you help on checking it or guide me to where I can learn to do it.

rmetzger commented:
Note: the bot is not reporting it, but the e2e tests of this PR are failing on azure: https://dev.azure.com/rmetzger/Flink/_build/results?buildId=6032&view=results

TisonKun commented:
@rmetzger Thanks for your notification! There are several homogenous failure due to

1. We reverse the dependencies from flink-streaming-java to flink-client.
2. Modules that contain failing tests depend on only flink-streaming-java.
3. However, these modules relied on filnk-clients for Executor service loading things.

Thus here is where, as described in the mailing list, we should explicitly add the dependency to flink-clients.

I previously introduce the reverse commits in a codebase without Executor service loading logics so it might be the case that we let the test spot where we should update dependencies.

TisonKun commented:
Latest Azure build passed: https://dev.azure.com/rmetzger/Flink/_build/results?buildId=6040&view=results

aljoscha commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/generator/JobGraphGeneratorUtils.java:
```diff
@@ -0,0 +1,72 @@
+/**
+ * Utilities for generating {@link JobGraph}.
+ */
+public enum JobGraphGeneratorUtils {
```
Why is this not `JobGraphUtils`. Or maybe even a method on `JobGraph`?

TisonKun replied:
I agree with `JobGraphUtils`. Since it includes logics packaging files I'd like to hold it in a separated place since `JobGraph` is generally regarded as a data class.

aljoscha commented on flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java:
```diff
@@ -128,16 +128,23 @@ public static void executeProgram(
 			LOG.info("Starting program (detached: {})", !configuration.getBoolean(DeploymentOptions.ATTACHED));
 
-			ContextEnvironmentFactory factory = new ContextEnvironmentFactory(
+			ContextEnvironment benv = new ContextEnvironment(
```
This changes the behaviour quite a bit? Before, every call to `ExecutionEnvironment.getEnvironment()` would create a new env from the factory. Now we only have one environment and keep returning that from the anonymous inner factory created in `setAsContext()`. Whats the reason for this change?

TisonKun replied:
I forget user could call ExecutionEnvironment multiple times. Generally speaking these envs have same configuration but if we then modify them in user code they should changed.

The motivation I do this change is from `OptimizerPlanEnvironment` where return identical instance of env for all `getEnvironment` calls. However, `StreamPlanEnvironment` is different instance.

As the semantic of `getEnvironment`, I think return different instance for each call is reasonable.

Will update.

TisonKun replied:
Updated.

aljoscha commented on flink-clients/src/main/java/org/apache/flink/client/FlinkPipelineTranslationUtil.java:
```diff
@@ -54,45 +54,19 @@ public static String translateToJSONExecutionPlan(Pipeline pipeline) {
-		return streamGraphTranslator;
-	}
-
-	private static FlinkPipelineTranslator reflectStreamGraphTranslator() {
```
Nice, that we don't need this additional reflection anymore. 😃

TisonKun commented:
@aljoscha Thanks for your review! I've addressed your comments inline.

For moving test to `flink-tests`, I agree that throwing everything into a full dependencies test module is bad practice. However, these tests don't belong to `flink-client` because they are primarily tests for streaming api. It happens we need to load services that are in `flink-client`. See also ExecutionEnvironment tests in `flink-tests`.

If we want to clear dependencies for testing, and for the long term maintenance, I'd prefer re-struct `flink-tests`, `flink-yarn-tests`, `flink-fs-tests` to be under a parent module and separate these env related tests in a sub module.

aljoscha commented on flink-clients/src/main/java/org/apache/flink/client/program/OptimizerPlanEnvironment.java:
```diff
@@ -22,143 +22,39 @@
 public class OptimizerPlanEnvironment extends ExecutionEnvironment {
 
-	private Pipeline pipeline;
+	private static Pipeline pipeline;
```
This change has some very subtle implications. Before, `OptimizerPlanEnvironment` was thread save, because `initializeContextEnvironment()` would store the environment factory in a thread-local field and the `Pipeline` field was not static. Now, multiple threads would try and store the pipeline in this one shared static field.

TisonKun replied:
What if users write code

```java
ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();

parallel-n {
  env.execute();
}
```

In this case, the whole function is not thread transparent. And even with master branch we might miss the exception. Here, already we make some assumption that user calls `env.execute` in the same thread he runs `main` and doesn't caught `ProgramAbortException`.

If we keep `OptimizerPlanEnvironment` as is, then we go into the problem you mention above for ContextEnvironment. What if

```java
ExecutionEnvironment env1 = ExecutionEnvironment.getExecutionEnvironment();
ExecutionEnvironment env2 = ExecutionEnvironment.getExecutionEnvironment();

env1 ...
env2 ...

env1.execute()
```

Actually `env1` & `env2` is the same environment.

Either way, we make some implications. And if you compare master `OptimizerPlanEnvironment` and `StreamPlanEnvironment`, you will find that they are different.

- `OptimizerPlanEnvironment.setAsContext` always set the same instance
- `StreamPlanEnvironment.setAsContext` always set a new instance, though it finally set StreamGraph to the single `OptimizerPlanEnvironment`, but different `StreamPlanEnvironment` instance has different configuration and the first one call `execute()` wins. If you put it in multi-threaded environment, the one runs as the same thread of `main` and first call `execute()` wins.

TisonKun replied:
It is, unfortunately, a historical problem that becomes obvious in this pull request. And anyway, we can tweak the code a bit to keep the manner as is. For the specific case, we use single `OptimizerPlanEnvironment` and `ContextEnvironmentFactory` like `StreamPlanEnvironment`.

TisonKun commented:
After a discussion with @aljoscha , let pipeline become `static` will fail on concurrent `JarRunHandler` case. We turn out to make `*PlanEnvironment` return `this` as a factory. This has several implication include

Multiple calls of `Env.getEnv` actually return the same instance, so users are recommended avoid calling `getEnv` multiple times.
